roofit/roofitcore/src/RooDataHist.cxx

```diff
@@ -1142,6 +1142,49 @@ RooPlot *RooDataHist::plotOn(RooPlot *frame, PlotOpt o) const
   return RooAbsData::plotOn(frame,o) ;
 }
 
+
+void RooDataHist::weights(double* output, RooSpan<double const> xVals, bool correctForBinSize) 
+{
+  auto nEvents = xVals.size();
+
+  RooAbsBinning const& binning = *_lvbins[0];
+
+  if (binning.isUniform()) {
+    double minVal = binning.binLow(0);
+    double maxValFirstBin = binning.binHigh(0);
+
+    double oneOverBinWidth;
+    if (correctForBinSize) {
+      oneOverBinWidth = 1. / (maxValFirstBin - minVal);
+    } else {
+      oneOverBinWidth = 1.;
+    }
+
+    for(std::size_t i = 0; i < nEvents; ++i) {
+      auto binIndex = static_cast<int>((xVals[i] - minVal) * oneOverBinWidth);
+      output[i] = weight(binIndex) * oneOverBinWidth;
+    }
+  }
+  else {
+    std::vector<int> binIndices(nEvents);
+    binning.binNumbers(xVals.data(), binIndices.data(), nEvents);
+
+    if(correctForBinSize) {
+      double* boundaries = binning.array();
+      for (std::size_t i=0; i < nEvents; ++i) {
+        auto binWidth = boundaries[binIndices[i] + 1] - boundaries[binIndices[i]];
+        output[i] = weight(binIndices[i]) / binWidth;
+      }
```

Comment:
```suggestion
      for (std::size_t i=0; i < nEvents; ++i) {
        int idx = binIndices[i];
        output[i] = _wgt[i] / _binv[i];
      }
```
I just remembered that there is a private member of the RooDataHist that caches the bin volumes. Can you try to use that? Also, I'd use the `_wgt` data member directly instead of the weight function, to make the code more explicit (`_wgt` is private, but not that the code is moved to RooDataHist this is not a problem anymore). The same applies to the other calls to `weight()` in the other code branches.